system_metrics/basic_sender.go

```diff
@@ -0,0 +1,202 @@
+package systemmetrics
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/evergreen-ci/timber"
+	"github.com/evergreen-ci/timber/internal"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+)
+
+// CompressionType describes how the system metrics data is compressed.
+type CompressionType int32
+
+// Valid CompressionType values.
+const (
+	CompressionTypeNone  CompressionType = 0
+	CompressionTypeTARGZ CompressionType = 1
+	CompressionTypeZIP   CompressionType = 2
+	CompressionTypeGZ    CompressionType = 3
+	CompressionTypeXZ    CompressionType = 4
+)
+
+func (f CompressionType) validate() error {
+	switch f {
+	case CompressionTypeNone, CompressionTypeTARGZ, CompressionTypeZIP, CompressionTypeGZ, CompressionTypeXZ:
+		return nil
+	default:
+		return errors.New("invalid compression type specified")
+	}
+}
+
+// SchemaType describes how the time series data is stored.
+type SchemaType int32
+
+// Valid SchemaType values.
+const (
+	SchemaTypeRawEvents             SchemaType = 0
+	SchemaTypeCollapsedEvents       SchemaType = 1
+	SchemaTypeIntervalSummarization SchemaType = 2
+	SchemaTypeHistogram             SchemaType = 3
+)
+
+func (f SchemaType) validate() error {
+	switch f {
+	case SchemaTypeRawEvents, SchemaTypeCollapsedEvents, SchemaTypeIntervalSummarization, SchemaTypeHistogram:
+		return nil
+	default:
+		return errors.New("invalid schema type specified")
+	}
+}
+
+type SystemMetricsClient struct {
+	client     internal.CedarSystemMetricsClient
+	clientConn *grpc.ClientConn
+}
+
+type ConnectionOptions struct {
+	DialOpts timber.DialCedarOptions
+	Client   http.Client
+}
+
+// NewSystemMetricsClient returns a grpc client to send system metrics data to
+// cedar. If authentication credentials (username and apikey) are not specified,
+// then an insecure connection will be established with the specified address
+// and port.
+func NewSystemMetricsClient(ctx context.Context, opts ConnectionOptions) (*SystemMetricsClient, error) {
+	var conn *grpc.ClientConn
+	var err error
+
+	if opts.DialOpts.APIKey == "" || opts.DialOpts.Username == "" {
```

Comment:
This should probably have a separate validate func on the opts struct that checks if both are set (it should be invalid that one is set and not the other) and also checks for that base address and port are set if username and api key are not.